### ISSUE TITLE
Arm feature detection: Never detecting neon on Raspberry pi 4

### ISSUE THREAD

**Issue by bluss** [rust-lang/stdarch#966]
Testing on an rpi 4 (64-bit ARM) with 32-bit userland target armv7-unknown-linux-gnueabihf (common rpi configuration).

`is_arm_feature_detected!("neon")` always returns false.

Test program `test.rs`:

```rust
#![feature(stdsimd)]
fn main() {
    dbg!(is_arm_feature_detected!("neon"));
    dbg!(cfg!(target_feature = "neon"));
    dbg!(is_arm_feature_detected!("crc"));
    dbg!(cfg!(target_feature = "crc"));
}
```

Running the program, detection reports false:

```rust
$ rustc +nightly test.rs  && ./test
[test.rs:5] is_arm_feature_detected!("neon") = false
[test.rs:6] cfg!(target_feature = "neon") = false
[test.rs:7] is_arm_feature_detected!("crc") = false
[test.rs:8] cfg!(target_feature = "crc") = false

$ rustc +nightly -Ctarget-feature=+neon test.rs  && ./test
[test.rs:5] is_arm_feature_detected!("neon") = false
[test.rs:6] cfg!(target_feature = "neon") = true
[test.rs:7] is_arm_feature_detected!("crc") = false
[test.rs:8] cfg!(target_feature = "crc") = false

$ rustc +nightly -Ctarget-feature=+neon,+crc test.rs  && ./test
[test.rs:5] is_arm_feature_detected!("neon") = false
[test.rs:6] cfg!(target_feature = "neon") = true
[test.rs:7] is_arm_feature_detected!("crc") = false
[test.rs:8] cfg!(target_feature = "crc") = true
```

The auxvec should report neon in this case, looks like it does:

```
$ LD_SHOW_AUXV=1 cat /dev/null
AT_SYSINFO_EHDR: 0xbefe1000
AT_HWCAP:        half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm
...
```

And for intrinsics like vaddq_f32 it does emit the expected add instruction.

Rust version: rustc 1.50.0-nightly (0f6f2d681 2020-12-06)

**Comment by Amanieu** [comment]
It seems that std_detect only reads AT_HWCAP or /proc/cpuinfo if the `use_std` feature is enabled, but no code enables this. Not that this would work in any case since `std_detect` can't depend on `std`.

I think the feature detection code needs to be rewritten to use `libc` directly instead of `std`, at which point we can unconditionally enable it.

**Comment by bluss** [comment]
Shouldn't `cfg!(target_feature = "neon")` also override the feature detection and make it flip to true? Or is that just an old behaviour?

**Comment by Lokathor** [comment]
I believe that the rust target in question doesn't globally enable `neon` at compile time.

**Comment by bluss** [comment]
std_detect_dlsym_getauxval looks like it's already using just libc, that code just needs to be enabled. From the looks of it, there is a nest of used code here that won't work if it is enabled - like code inside `#[cfg(not(feature = "std_detect_dlsym_getauxval"))]` calling `ffi_getauxval` with no explanation of where the function ffi_getauxval comes from (it has not been declared).

@Lokathor I can't quite make out what that means in relation to this issue

**Comment by Lokathor** [comment]
`cfg!(target_feature = "neon")` evaluates at compile time, so if `neon` is not enabled at compile time it will evaluate to false. And I think that the target you named in the opening post (`armv7-unknown-linux-gnueabihf`) doesn't enable `neon` by default.

You could turn on `neon` at compile time via `RUSTFLAGS` and using `-Ctarget-feature=+neon` or (probably) with `-Ctarget-cpu=native`.

**Comment by Lokathor** [comment]
Oh my, you did do that now that i see some of your test code lines more closely. Ignore me.

**Comment by bluss** [comment]
Maybe this was fixed by #1005, which would be awesome?

**Comment by bluss** [comment]
nowadays fixed :slightly_smiling_face:
